### PR TITLE
#170947687 User should be able to like/unlike accommodation

### PR DESCRIPTION
**Pull request by EugeneMunya**
#### What does this PR do?
- Add functionality for allowing a traveler to like or unlike a specific accommodation
#### Description of Task to be completed?
- User can like or dislike a specific accommodation
#### How should this be manually tested?
- open your terminal and run this command `git clone https://github.com/andela/knightsbn   backend.git` to clone repository
- run `cd knights-bn-backend` to get inside the project
- run `git fetch && git checkout ft-like-unlike-accommodation-170947687` 
- run `npm install` to install dependencies
- check if you have PostgreSQL database if you don't have it install it from here [PostgreSQL](https://www.2ndquadrant.com/en/blog/pginstaller-install-postgresql/)
- create database
- create `.env` file in your project root directory
- follow `env.example` file to create environment variables in `.env` file and 
   assign them with all your database configurations
- run `npm run dropTables` to clean your database
- run `npm run createTables` to create tables in your new database you have created.
- run `npm run dev-start` to start the server.
- open your postman or any other API tester you have
- you can log in with default credentials below
```
{
    "email": "alain.maxime@gmail.com",
    "password": "Niyonkuru@1"
}
```
- view all available accommodation facilities by sending a GET request to this URL `http://localhost:4000/api/v1/view/accommodations`  in order to find an accommodation you want to like or dislike
- chose the one you want to like or dislike by grabbing its ID and append it at the end of this URL `http://localhost:4000/api/v1/accommodation/dislike/1` or this one `http://localhost:4000/api/v1/accommodation/like/1` and sent POST request to it
- to view, all the accommodations with their likes and dislikes sent GET request to this URL `http://localhost:4000/api/v1/view/accommodations`
- to view specific accommodation along with its likes and dislikes send a GET request to this URL `http://localhost:4000/api/v1/view/accommodation/1` the number one `(1)` you see here is accommodation `ID`.  you can replace it with another accommodation `ID`  you want to view

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#170947687](https://www.pivotaltracker.com/story/show/170947687)
#### perform like
Screen Shot 2020-03-19 at 17 39 58
#### perform dislike
Screen Shot 2020-03-19 at 17 39 03

#### view likes and dislikes of a specific accommodation 
Screen Shot 2020-03-19 at 17 41 19

### CODE REVIEW

**Comment by T2Wil** [comment]
@EugeneMunya , remember to integrate the two commits into one.

**Comment by EugeneMunya** [comment]
> @EugeneMunya , remember to integrate the two commits into one.

@T2Wil I am about to resolve that issue

**Comment by T2Wil** [comment]
@EugeneMunya , you did a wonderfully job here, now I can like and dislike an accommodation as it is expected. 
My problem now is why are we having the first index set to `0` in the arrays of  `likesUsers` and `dislikeUsers`?

image

File: README.md [hunk 1]
```diff
@@ -114,6 +114,12 @@ Open terminal from your computer
 
 `- GET /trips/stats/:date? - For a manager and a user to get statistics of the trips made.`
 
+`- GET /accommodation/like/:id? - For liking accommodation.`
+
+`- GET /accommodation/dislike/:id? - For disliking accommodation.`
+
+/accommodation/like/:id
```
Comment (T2Wil) [inline README.md]: This line should be removed.

File: src/controllers/accommodation.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import { config } from 'dotenv';
 import _ from 'lodash';
 import models, { Accommodation, Bookings } from '../db/models';
-import Accomodation from '../helpers/queries';
+import query from '../helpers/queries';
```
Comment (T2Wil) [inline src/controllers/accommodation.js]: Why did you rename `Accomodation` to `query`?
I think `Accommodation` is a good name to give to a class of queries related to accommodations.

Comment (EugeneMunya) [reply src/controllers/accommodation.js]: It has the same name as our table name

File: README.md (hunk 1 above)
Comment (EugeneMunya) [reply README.md]: Let me remove it

**Comment by EugeneMunya** [comment]
> @EugeneMunya , you did a wonderfully job here, now I can like and dislike an accommodation as it is expected.
> My problem now is why are we having the first index set to `0` in the arrays of `likesUsers` and `dislikeUsers`?
> 
> image

when there is no initial value, sequelize doesn't know that the field is an array, and you can't apply forEach on that field. to solve that problem I decided to initialize the array by `0` you can see in Accommodation migration these fields have default value of `0`

File: src/controllers/accommodation.js (hunk 2 above)
Comment (T2Wil) [reply src/controllers/accommodation.js]: Having the same name as table does it affect anything?

**Comment by Niyonsengaeric** [comment]
@EugeneMunya  have you seen that the maintainability is failing??

**Comment by EugeneMunya** [comment]
> @EugeneMunya have you seen that the maintainability is failing??

@Niyonsengaeric I have seen it and I worked on it

**Review by T2Wil — APPROVED** [review]

File: src/controllers/accommodation.js [hunk 3]
```diff
@@ -163,5 +163,33 @@ export default class accomodationFacility {
...
     }
   }
+
+  static likeAccommodation(req, res) {
+    const accommId = req.params.id;
+    Accommodation
+      .findOne({ where: { id: accommId } }).then((accomm) => {
+        const { likes, likesUsers } = accomm;
+        likesUsers.push(req.user.id);
+        Accommodation.update({ likes: likes + 1, likesUsers },
```
Comment (kagabof) [inline src/controllers/accommodation.js]: What will happen if it doesn't update?

File: src/controllers/accommodation.js [hunk 4]
```diff
@@ -163,5 +163,33 @@ export default class accomodationFacility {
       return res.status(500).json({ status: 500, errorMessage: error });
     }
   }
+
+  static likeAccommodation(req, res) {
+    const accommId = req.params.id;
+    Accommodation
+      .findOne({ where: { id: accommId } }).then((accomm) => {
```
Comment (kagabof) [inline src/controllers/accommodation.js:170]: What will happen if it doesn't find an acommodation or it throws an error?

File: src/controllers/accommodation.js [hunk 5]
```diff
@@ -163,5 +163,33 @@ export default class accomodationFacility {
...
+      });
+  }
+
+  static async dislikeAccommodation(req, res) {
+    try {
+      const accommId = req.params.id;
+      const { dislikes, dislikeUsers } = await accommodation
+        .verifyAccom({ id: accommId }, Accommodation);
+      dislikeUsers.push(req.user.id);
+      await Accommodation
```
Comment (kagabof) [inline src/controllers/accommodation.js]: I think you need to check if it is have updated successfully.

File: src/helpers/likeAndDislike.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
+export const fromDisLikeToLike = (likesUsers, dislikeUsers, id) => {
+  dislikeUsers.forEach((userId, index) => {
+    if (userId === id) {
+      dislikeUsers.splice(index, 1);
```
Comment (kagabof) [inline src/helpers/likeAndDislike.js]: This is causing O(n^2) and it is very bad, why not change it into O(n).

File: src/helpers/likeAndDislike.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+  });
+  return likesUsers;
+};
+
+export const fromLikeToDislike = (usersDislike, usersLikes, id) => {
+  const updatedUsers = [];
+  usersLikes.forEach((uId, i) => {
+    if (uId === id) {
+      usersLikes.splice(i, 1);
+      usersDislike.push(id);
```
Comment (kagabof) [inline src/helpers/likeAndDislike.js]: This is causing O(n^2) and it is very bad, why not change it into O(n).

**Review by kagabof — CHANGES_REQUESTED** [review]

File: src/controllers/accommodation.js (hunk 4 above)
Comment (EugeneMunya) [reply src/controllers/accommodation.js]: There is a middleware checks for that

File: src/helpers/likeAndDislike.js (hunk 6 above)
Comment (EugeneMunya) [reply src/helpers/likeAndDislike.js]: > This is causing O(n^2) and it is very bad, why not change it into O(n).

@kagabof since my logic to that functionality has ended  you can give me an idea of how I can change into O(n)

File: src/controllers/accommodation.js (hunk 4 above)
Comment (EugeneMunya) [reply src/controllers/accommodation.js]: Screen Shot 2020-03-21 at 21 51 03

this a middleware that checks everything for that function

File: src/controllers/accommodation.js (hunk 3 above)
Comment (EugeneMunya) [reply src/controllers/accommodation.js]: since the update query depends on findOne query we believe that if findOne is working well the update it will work well too.

File: src/controllers/accommodation.js (hunk 4 above)
Comment (EugeneMunya) [reply src/controllers/accommodation.js:170]: I have managed to solve these problems

File: src/helpers/likeAndDislike.js (hunk 6 above)
Comment (EugeneMunya) [reply src/helpers/likeAndDislike.js]: I have tried to use the paths I could think of but none is helping me

File: src/helpers/likeAndDislike.js (hunk 7 above)
Comment (EugeneMunya) [reply src/helpers/likeAndDislike.js]: I see it bad but I can't find any other way I can use to solve that problem, If you have any you can share to me

**Review by kagabof — CHANGES_REQUESTED** [review]
Please check your commit message, there is an empty line missing.

**Comment by EugeneMunya** [comment]
> Please check your commit message, there is an empty line missing.

Let me edit it, its because the title of the commit is too long

**Review by kagabof — APPROVED** [review]
